src/data/nat/basic.lean

```diff
@@ -743,6 +743,21 @@ lemma decreasing_induction_succ_left {P : ℕ → Sort*} (h : ∀n, P (n+1) →
 by { rw [subsingleton.elim mn (le_trans (le_succ m) smn), decreasing_induction_trans,
          decreasing_induction_succ'] }
 
+/-- Recursion principle on even and odd numbers: if we have `P 0`, and for all `i : ℕ` we can
+extend from `P i` to both `P (2 * i)` and `P (2 * i + 1)`, then we have `P n` for all `n : ℕ`.
+This is nothing more than a wrapper around `nat.binary_rec`, to avoid having to switch to
+dealing with `bit0` and `bit1`. -/
+@[elab_as_eliminator]
+def even_odd_rec (n : ℕ) (P : ℕ → Sort*) (h0 : P 0)
```

Comment:
Unless the idea is that this is only to be used for propositions, you should prove the following:
```lean
@[simp] lemma even_odd_rec_zero : even_odd_rec 0 P h0 h_even h_odd = h0 := sorry
@[simp] lemma even_odd_rec_even :
  even_odd_rec (2 * n) P h0 h_even h_odd = h_even n (even_odd_rec n P h0 h_even h_odd) := sorry
@[simp] lemma even_odd_rec_odd : 
  even_odd_rec (2 * n + 1) P h0 h_even h_odd = h_odd n (even_odd_rec n P h0 h_even h_odd) := sorry
```
The latter two seem to be missing for `binary_rec` too, so it might be worth starting there.

Reply:
The first one is easy. I suspect the other two will require some `heq` tampering:
```lean
@[simp] lemma even_odd_rec_zero (P : ℕ → Sort*) (h0 : P 0)
  (h_even : ∀ i, P i → P (2 * i)) (h_odd : ∀ i, P i → P (2 * i + 1)) : 
  @even_odd_rec 0 P h0 h_even h_odd = h0 := binary_rec_zero _ _
```

Reply:
Actually, here you go:
```suggestion
@[simp] lemma even_odd_rec_even (n : ℕ) (P : ℕ → Sort*) (h0 : P 0)
  (h_even : ∀ i, P i → P (2 * i)) (h_odd : ∀ i, P i → P (2 * i + 1))
  (H : h_even 0 h0 = h0) :
  @even_odd_rec (2 * n) P h0 h_even h_odd = h_even n (even_odd_rec n P h0 h_even h_odd) :=
begin
  convert binary_rec_eq _ ff n,
  { exact (bit0_eq_two_mul _).symm },
  { exact (bit0_eq_two_mul _).symm },
  { apply heq_of_cast_eq,
    refl },
  { exact H }
end

@[simp] lemma even_odd_rec_odd (n : ℕ) (P : ℕ → Sort*) (h0 : P 0)
  (h_even : ∀ i, P i → P (2 * i)) (h_odd : ∀ i, P i → P (2 * i + 1))
  (H : h_even 0 h0 = h0) :
  @even_odd_rec (2 * n + 1) P h0 h_even h_odd = h_odd n (even_odd_rec n P h0 h_even h_odd) :=
begin
  convert binary_rec_eq _ tt n,
  { exact (bit0_eq_two_mul _).symm },
  { exact (bit0_eq_two_mul _).symm },
  { apply heq_of_cast_eq,
    refl },
  { exact H }
end
```

Reply:
Fantastic, thank you!